test_get_fqdn_by_host_on_bad_host fails when a VPN is in use

**Describe the bug**
In certain environments (such as when a VPN is in use) the IOError raised by socket.gethostbyname_ex contains a different error message than the one asserted in the test.

**Release version(s) and/or repository branch(es) affected?**
master branch

**Steps to reproduce the bug**
I encountered this error when running the test suite on MacOS 10.14.6 while connected to a Cisco VPN. It does not occur on the same system when the VPN is disconnected. It probably occurs in other environments as well, but I'm not sure which ones.

**Expected behavior**
The test should pass. In the environment I describe above it fails because socket.gethostbyname_ex throws an exception with the message "[Errno 8] nodename nor servname provided, or not known: 'nosuchhost.nosuchdomain.org'", and the test is looking for the message "[Errno -2] Name or service not known: 'nosuchhost.nosuchdomain.org'".

**Test output**
```
=================================== FAILURES ===================================
______________ TestHostUserUtil.test_get_fqdn_by_host_on_bad_host ______________
[gw0] darwin -- Python 3.8.6 /Users/jhaiduce/Development/cylc-flow/venv/bin/pyth
on3

self = <unit.test_hostuserutil.TestHostUserUtil testMethod=test_get_fqdn_by_host
_on_bad_host>

    def test_get_fqdn_by_host_on_bad_host(self):
        """get_fqdn_by_host bad host."""
        bad_host = 'nosuchhost.nosuchdomain.org'
        try:  # Future: Replace with assertRaises context manager syntax
>           get_fqdn_by_host(bad_host)

tests/unit/test_hostuserutil.py:44: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

target = 'nosuchhost.nosuchdomain.org'

    def get_fqdn_by_host(target):
        """Shorthand for HostUtil.get_inst().get_fqdn_by_host(target)."""
>       return HostUtil.get_inst().get_fqdn_by_host(target)

cylc/flow/hostuserutil.py:248: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cylc.flow.hostuserutil.HostUtil object at 0x1025f31c0>
target = 'nosuchhost.nosuchdomain.org'

    def get_fqdn_by_host(self, target):
        """Return the fully qualified domain name of the target host."""
        if not self.is_remote_host(target):
            return self.get_host()
>       return self._get_host_info(target)[0]

cylc/flow/hostuserutil.py:154: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cylc.flow.hostuserutil.HostUtil object at 0x1025f31c0>
target = 'nosuchhost.nosuchdomain.org'

    def _get_host_info(self, target=None):
        """Return the extended info of the current host."""
        if target not in self._host_exs:
            if target is None:
                target = socket.getfqdn()
            try:
>               self._host_exs[target] = socket.gethostbyname_ex(target)
E               socket.gaierror: [Errno 8] nodename nor servname provided, or not known: 'nosuchhost.nosuchdomain.org'

cylc/flow/hostuserutil.py:118: gaierror

During handling of the above exception, another exception occurred:

self = <unit.test_hostuserutil.TestHostUserUtil testMethod=test_get_fqdn_by_host_on_bad_host>

    def test_get_fqdn_by_host_on_bad_host(self):
        """get_fqdn_by_host bad host."""
        bad_host = 'nosuchhost.nosuchdomain.org'
        try:  # Future: Replace with assertRaises context manager syntax
            get_fqdn_by_host(bad_host)
        except IOError as exc:
            self.assertEqual(exc.filename, bad_host)
>           self.assertEqual(
                "[Errno -2] Name or service not known: '%s'" % bad_host,
                str(exc))
E           AssertionError: "[Errno -2] Name or service not known: 'nosuchhos[15 chars]org'" != "[Errno 8] nodename nor servname provided, or not[33 chars]org'"
E           - [Errno -2] Name or service not known: 'nosuchhost.nosuchdomain.org'
E           ?        ^^  ^            ^
E           + [Errno 8] nodename nor servname provided, or not known: 'nosuchhost.nosuchdomain.org'
E           ?        ^  ^^^^^    +       +++++++++ ^ +++++

tests/unit/test_hostuserutil.py:47: AssertionError
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -95,6 +95,8 @@ config option `[scheduling]stop after cycle point`.
 
 ### Fixes
 
+[#3917](https://github.com/cylc/cylc-flow/pull/3917) - Fix a bug that caused one of the hostname resolution tests to fail in certain environments.
+
 [#3879](https://github.com/cylc/cylc-flow/pull/3879) - Removed Google
 Groups e-mail from pip packaging metadata. Users browsing PYPI will have
 to visit our website to find out how to reach us (we are using Discourse

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -70,6 +70,7 @@ requests_).
  - Gilliano Menezes
  - Mel Hall
  - Ronnie Dutta
+ - John Haiducek
 
 (All contributors are identifiable with email addresses in the git version
 control logs or otherwise.)

--- a/tests/unit/test_hostuserutil.py
+++ b/tests/unit/test_hostuserutil.py
@@ -40,13 +40,12 @@ class TestHostUserUtil(unittest.TestCase):
     def test_get_fqdn_by_host_on_bad_host(self):
         """get_fqdn_by_host bad host."""
         bad_host = 'nosuchhost.nosuchdomain.org'
-        try:  # Future: Replace with assertRaises context manager syntax
+        with self.assertRaisesRegex(
+                IOError,
+                "(\[Errno -2\] Name or service|\[Errno 8\] nodename nor servname provided, or) not known: '{}'".format(bad_host)) as ctx:
             get_fqdn_by_host(bad_host)
-        except IOError as exc:
-            self.assertEqual(exc.filename, bad_host)
-            self.assertEqual(
-                "[Errno -2] Name or service not known: '%s'" % bad_host,
-                str(exc))
+
+        self.assertEqual(ctx.exception.filename, bad_host)
 
     def test_get_user(self):
         """get_user."""

--- a/tests/unit/test_hostuserutil.py
+++ b/tests/unit/test_hostuserutil.py
@@ -42,7 +42,9 @@ class TestHostUserUtil(unittest.TestCase):
         bad_host = 'nosuchhost.nosuchdomain.org'
         with self.assertRaisesRegex(
                 IOError,
-                "(\[Errno -2\] Name or service|\[Errno 8\] nodename nor servname provided, or) not known: '{}'".format(bad_host)) as ctx:
+                r"(\[Errno -2\] Name or service|"
+                r"\[Errno 8\] nodename nor servname provided, or)"
+                r" not known: '{}'".format(bad_host)) as ctx:
             get_fqdn_by_host(bad_host)
 
         self.assertEqual(ctx.exception.filename, bad_host)

